pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,295 @@
+	imc.Status.InitializeConditions()
+
+	if imc.DeletionTimestamp != nil {
+		// Everything is cleaned up by the garbage collector.
+		return nil
+	}
+
+	// We reconcile the status of the Channel by looking at:
+	// 1. Dispatcher Deployment for it's readiness.
+	// 2. Dispatcher k8s Service for it's existence.
+	// 3. Dispatcher endpoints to ensure that there's something backing the Service.
+
+	// Get the Dispatcher Deployment and propagate the status to the Channel
+	d, err := r.deploymentLister.Deployments(r.dispatcherNamespace).Get(r.dispatcherDeploymentName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			imc.Status.MarkDispatcherFailed("DispatcherDeploymentDoesNotExist", "Dispatcher Deployment does not exist")
+		} else {
+			logging.FromContext(ctx).Error("Unable to get the dispatcher Deployment", zap.Error(err))
+			imc.Status.MarkDispatcherFailed("DispatcherDeploymentGetFailed", "Failed to get dispatcher Deployment")
+		}
+		return err
+	}
+	imc.Status.PropagateDispatcherStatus(&d.Status)
+
+	// Get the Dispatcher Service and propagate the status to the Channel in case it does not exist.
+	// We don't do anything with the service because it's status contains nothing useful, so just do
+	// an existence check. Then below we check the endpoints targeting it.
+	_, err = r.serviceLister.Services(r.dispatcherNamespace).Get(r.dispatcherServiceName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			imc.Status.MarkServiceFailed("DispatcherServiceDoesNotExist", "Dispatcher Service does not exist")
+		} else {
+			logging.FromContext(ctx).Error("Unable to get the dispatcher service", zap.Error(err))
+			imc.Status.MarkServiceFailed("DispatcherServiceGetFailed", "Failed to get dispatcher service")
+		}
+		return err
+	}
+
+	imc.Status.MarkServiceTrue()
+
+	// Get the Dispatcher Service Endpoints and propagate the status to the Channel
+	// endpoints has the same name as the service, so not a bug.
+	e, err := r.endpointsLister.Endpoints(r.dispatcherNamespace).Get(r.dispatcherServiceName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			imc.Status.MarkEndpointsFailed("DispatcherEndpointsDoesNotExist", "Dispatcher Endpoints does not exist")
+		} else {
+			logging.FromContext(ctx).Error("Unable to get the dispatcher endpoints", zap.Error(err))
+			imc.Status.MarkEndpointsFailed("DispatcherEndpointsGetFailed", "Failed to get dispatcher endpoints")
+		}
+		return err
+	}
+
+	if len(e.Subsets) == 0 {
+		logging.FromContext(ctx).Error("No endpoints found for Dispatcher service", zap.Error(err))
+		imc.Status.MarkEndpointsFailed("DispatcherEndpointsNotReady", "There are no endpoints ready for Dispatcher service")
+		return errors.New("there are no endpoints ready for Dispatcher service")
+	}
+
+	imc.Status.MarkEndpointsTrue()
+
+	imc.Status.SetAddress(fmt.Sprintf("%s.%s.svc.%s", r.dispatcherServiceName, r.dispatcherNamespace, utils.GetClusterDomainName()))
```

Comment:
Shouldn't it create a ExternalName K8s service, point the external name to the dispatcher service, and use the newly created  K8s as the host address? Unless you are planning to change the way dispatching happens?